kafka/src/main/scala/filodb/kafka/KafkaIngestionStream.scala

```diff
@@ -48,7 +54,20 @@ class KafkaIngestionStream(config: Config,
    */
   override def get: Observable[SomeData] =
     consumer.map { record =>
-      SomeData(record.value.asInstanceOf[RecordContainer], record.offset)
+      val rc = record.value.asInstanceOf[RecordContainer]
+
+      // We want to maintain the illusion that ingestion time never moves backwards. Because
+      // the record containers come from different sources, they might have clocks which are
+      // slightly skewed. Always choose the highest one, for simplicity and consistency.
+      highestTimestamp = rc.monotonicTimestamp(highestTimestamp)
+
+      // Note that if the lag is negative, report it as such. This makes it possible to observe
+      // cases where this server has a clock set to the past, or a source has a clock set to
+      // the future.
+      val lag = System.currentTimeMillis() - highestTimestamp
```

Comment:
What is the latency of each `currentTimeMillis` call, just curious?

Reply:
Also, we probably don't need to set the lag with every container.  Once a second is enough - consider moving this to the TimeSeriesShard.

Actually the more I think about it the more this logic should be in the TimeSeriesShard, not here in Kafka, as it is universal for all ingestion streams, not just Kafka.

Reply:
Getting the current time is relatively cheap. It doesn't involve full system call overhead. Considering that the check is performed on the container, it's not called that much overall.

Reply:
Ideally, the time check should be global, and performing a per-shard adjustment means that the shards can be out sync with each other.